typing: storage

aorenste opened the pull request:
This isn't a full typing of the file - it just fixes some uses of unbound 'T' (if you use a TypeVar as an output it also needs to be an input).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130670
* __->__ #130669

aorenste committed:
typing: storage

[ghstack-poisoned]

Skylion007 commented on torch/storage.py:
```diff
@@ -1,11 +1,15 @@
 # mypy: allow-untyped-defs
+
+from __future__ import annotations
```
Why do we need to turn on future annotations here? Any particular reason?

aorenste replied:
Sure - quoting works fine. But we could ask the opposite - what's the downside to turning it on?

Skylion007 replied:
New PEP is considering phasing it out.

aorenste replied:
I assume you're talking about PEP749? From what I can tell it's just phasing out the import statement, not the deferred evaluation aspect that it provides - so sure we'd have to strip the import statement in something like 5 years from now - but the behavior should still be the same.

Skylion007 commented on torch/storage.py:
```diff
@@ -852,12 +862,15 @@ def untyped(self):
+    def _new_wrapped_storage(self, untyped_storage) -> Self:
         assert type(untyped_storage) == torch.UntypedStorage
 
         if type(self) == TypedStorage:
```
```suggestion
        if type(self) is TypedStorage:
```

aorenste committed:
Update on "typing: storage"

This isn't a full typing of the file - it just fixes some uses of unbound 'T' (if you use a TypeVar as an output it also needs to be an input).




[ghstack-poisoned]

XuehaiPan commented on torch/storage.py:
```diff
@@ -57,10 +59,14 @@ def nbytes(self) -> _int:
         return _type(self, dtype, non_blocking)
 
-    def cuda(self, device=None, non_blocking=False) -> T:  # type: ignore[type-var, misc] # noqa: E704
+    def cuda(
+        self, device=None, non_blocking=False
+    ) -> Union[_StorageBase, TypedStorage]:
```
Add `from torch.types import Device`

```suggestion
    def cuda(
        self, device: Device = None, non_blocking: _bool = False
    ) -> Union[_StorageBase, TypedStorage]:
```

The argument should also be typed.

aorenste replied:
Is this correct (`Device`)? This function passes that argument to `torch.device` which is defined in Device.cpp (THPDevice_pynew) as taking `int64_t?` but `Device` is defined as `str | int | torch.device`.

XuehaiPan commented on torch/storage.py:
```diff
@@ -1,11 +1,13 @@
 # mypy: allow-untyped-defs
```
We should type all methods and remove this comment.

aorenste replied:
Totally agree - but that's a separate (and much larger) change (which I'm doing in a background task).  This is just removing some stuff that mypy actively complains about.

aorenste committed:
Update on "typing: storage"

This isn't a full typing of the file - it just fixes some uses of unbound 'T' (if you use a TypeVar as an output it also needs to be an input).




[ghstack-poisoned]
